rust/sbd-server/src/cslot.rs

```diff
@@ -0,0 +1,416 @@
+use std::collections::HashMap;
+use std::sync::{Arc, Mutex, Weak};
+
+static U: std::sync::atomic::AtomicU64 = std::sync::atomic::AtomicU64::new(1);
+
+enum TaskMsg {
+    NewWs {
+        uniq: u64,
+        index: usize,
+        ws: Arc<ws::WebSocket<MaybeTlsStream>>,
+        ip: Arc<std::net::Ipv6Addr>,
+        pk: PubKey,
+    },
+    Close,
+}
+
+struct SlotEntry {
+    send: tokio::sync::mpsc::UnboundedSender<TaskMsg>,
+}
+
+struct SlabEntry {
+    uniq: u64,
+    handshake_complete: bool,
+    weak_ws: Weak<ws::WebSocket<MaybeTlsStream>>,
+}
+
+struct CSlotInner {
+    max_count: usize,
+    slots: Vec<SlotEntry>,
+    slab: slab::Slab<SlabEntry>,
+    pk_to_index: HashMap<PubKey, usize>,
+    ip_to_index: HashMap<Arc<std::net::Ipv6Addr>, Vec<usize>>,
+    task_list: Vec<tokio::task::JoinHandle<()>>,
+}
+
+impl Drop for CSlotInner {
+    fn drop(&mut self) {
+        for task in self.task_list.iter() {
+            task.abort();
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct WeakCSlot(Weak<Mutex<CSlotInner>>);
+
+impl WeakCSlot {
+    pub fn upgrade(&self) -> Option<CSlot> {
+        self.0.upgrade().map(CSlot)
+    }
+}
+
+pub struct CSlot(Arc<Mutex<CSlotInner>>);
+
+impl CSlot {
+    pub fn new(config: Arc<Config>, ip_rate: Arc<ip_rate::IpRate>) -> Self {
+        let count = config.limit_clients as usize;
+        Self(Arc::new_cyclic(|this| {
+            let mut slots = Vec::with_capacity(count);
+            let mut task_list = Vec::with_capacity(count);
+            for _ in 0..count {
+                let (send, recv) = tokio::sync::mpsc::unbounded_channel();
+                slots.push(SlotEntry { send });
+                task_list.push(tokio::task::spawn(top_task(
+                    config.clone(),
+                    ip_rate.clone(),
+                    WeakCSlot(this.clone()),
+                    recv,
+                )));
+            }
+            Mutex::new(CSlotInner {
+                max_count: count,
+                slots,
+                slab: slab::Slab::with_capacity(count),
+                pk_to_index: HashMap::with_capacity(count),
+                ip_to_index: HashMap::with_capacity(count),
+                task_list,
+            })
+        }))
+    }
+
+    pub fn weak(&self) -> WeakCSlot {
+        WeakCSlot(Arc::downgrade(&self.0))
+    }
+
+    fn remove(&self, uniq: u64, index: usize) {
+        let mut lock = self.0.lock().unwrap();
+
+        match lock.slab.get(index) {
+            None => return,
+            Some(s) => {
+                if s.uniq != uniq {
+                    return;
+                }
+            }
+        }
+
+        let _ = lock.slots.get(index).unwrap().send.send(TaskMsg::Close);
+        lock.slab.remove(index);
+        lock.pk_to_index.retain(|_, i| *i != index);
+        lock.ip_to_index.retain(|_, v| {
+            v.retain(|i| *i != index);
+            !v.is_empty()
+        });
+    }
+
+    // oi clippy, this is super straight forward...
```

Comment:
:laughing: It has a pretty harsh limit on generics depth I think. I imagine the return type is bothering it